homeworks/solutions/torantie/3-smooth/src/halfedge.js

```diff
@@ -0,0 +1,632 @@
+        }
+
+        // set the current halfedge's attributes
+        bhe.vertex = next.vertex
+        bhe.edge = current.edge
+        bhe.onBoundary = true
+
+        // point the new halfedge and face to each other
+        bhe.face = f
+        f.halfedge = bhe
+
+        // point the new halfedge and twin to each other
+        bhe.twin = current
+        current.twin = bhe
+
+        current = next
+      } while(current != he)
+
+      // link the cycle of boundary halfedges together
+      const n = bcycle.length
+      for (let j = 0; j < n; j++) {
+        bcycle[j].next = bcycle[(j+n-1)%n]
+        bcycle[j].prev = bcycle[(j+1)%n]
+        hasTwin.set(bcycle[j], true)
+        hasTwin.set(bcycle[j].twin, true)
+      }
+    }
+
+    // reset indices
+    let index = 0
+    this.vertices.forEach(v => { v.idx = index++ })
+    index = 0
+    this.edges.forEach(e => { e.idx = index++ })
+    index = 0
+    this.faces.forEach(f => { f.idx = index++ })
+    index = 0
+    this.halfedges.forEach(h => { h.idx = index++ })
+
+    //Create Copy
+    for (const vertex of this.vertices) {
+      let vertexCopy = new Vertex()
+      vertexCopy.idx = vertex.idx
+      vertexCopy.halfedge = vertex.halfedge
+      vertexCopy.position = new Vector(vertex.position.x,vertex.position.y,vertex.position.z)
+      this.vertsOrig.push(vertexCopy)
+    }
+  }
+
+
+  laplaceMatrix(weightType) {
+    // TODO: construct the Laplace matrix.
+    const numberOfVertices = this.vertices.length
+    let weightTriplet = new Triplet(numberOfVertices, numberOfVertices)
+
+    for (const vert of this.vertices) {
+      const i = vert.idx
+      let sum = 1e-8 // Tikhonov regularization to get strict positive definite
+
+      vert.forEachHalfEdge(h => {
+        let w = 0
+
+        switch (weightType) {
+          case 'uniform':
+            w = 1
+            break
+          case 'cotan':
+            w = (h.cotan() + h.twin.cotan())/2
+        }
+
+        sum += w
+        //console.log("Value: "+w+" Position: ("+ i+","+h.twin.vertex.idx +")")
+        weightTriplet.addEntry(w, i, h.twin.vertex.idx)
+      })
+      weightTriplet.addEntry(-sum, i, i)
+      //console.log("Value: "+-sum+" Position: ("+ i+","+i +")")
+    }
+
+    let weightMatrix = SparseMatrix.fromTriplet(weightTriplet)
+
+    return weightMatrix
+  }
+
+  massMatrix(weightType){
+    const numberOfVertices = this.vertices.length
+    let massTriplet = new Triplet(numberOfVertices, numberOfVertices)
+
+    for (const vert of this.vertices) {
+      const i = vert.idx
+
+      switch (weightType) {
+        case 'uniform':
+          let neighbours = 0
+          vert.forEachHalfEdge(() => {
+            neighbours++
+          })
+
+          massTriplet.addEntry(neighbours,i,i)
+          break
+        case 'cotan':
+
+          let area = vert.calculateVoronoiArea()
+          // area alone is very small and destroys the smoothing
+          massTriplet.addEntry(1+area,i,i)
```

Comment:
This is the key to how the mass matrix could influence the linear solver.
Use `w` directly will result in a theoretically correct result: with a large time step (>0.1), changes to the mesh becomes large.

There are two ways to resolve this type of numerical issue:

1. Scale w using a constant
2. Just use identity matrix: Because almost all `w` are very small (the mesh itself is small), adding `1` here roughly makes the mass matrix become an identity matrix.